src/util.cc

```diff
@@ -1,8 +1,196 @@
 #include "util.h"
 #include "string_bytes.h"
 
+#define UNI_SUR_HIGH_START    (uint32_t) 0xD800
+#define UNI_SUR_LOW_END       (uint32_t) 0xDFFF
+#define UNI_REPLACEMENT_CHAR  (uint32_t) 0x0000FFFD
+#define UNI_MAX_LEGAL_UTF32   (uint32_t) 0x0010FFFF
+
+#if __GNUC__ >= 4 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 4)
+#define HAS_GCC_BUILTIN_CLZ
+#endif
+
+#ifdef WIN32
+#include <intrin.h>
+static uint32_t __inline clz(uint32_t xs) {
+  unsigned long result = 0;
+  _BitScanReverse(&result, xs);
+  return (31 - result);
+}
+#elif defined(HAS_GCC_BUILTIN_CLZ)
+#define clz(xs) __builtin_clz(xs)
+#else
+static inline uint32_t clz(uint32_t xs) {
+  uint32_t out = 0;
+  while (xs >> (31 - out)) {
+    ++out;
+  }
+  return out;
+}
+#endif
```

Comment:
with `__builtin_clz`:

```asm
	movl	%ebx, %eax
	shll	$24, %eax
	notl	%eax
	bsrl	%eax, %ecx    # <-- __builtin_clz (bitscanreverse)
	xorl	$31, %ecx
```

With `HAS_GCC_BUILTIN_CLZ` disabled manually:

```asm
	##DEBUG_VALUE: glyph <- R10D
	movl	%r10d, %ebx
	shll	$24, %ebx
	notl	%ebx
Ltmp17:
	##DEBUG_VALUE: clz:out <- 0
	##DEBUG_VALUE: clz:xs <- EBX
	xorl	%eax, %eax
Ltmp18:
LBB0_4:                                 ##   Parent Loop BB0_3 Depth=1
                                        ## =>  This Inner Loop Header: Depth=2
	##DEBUG_VALUE: StripInvalidUtf8Glyphs:input <- R14
	##DEBUG_VALUE: StripInvalidUtf8Glyphs:size <- R15
	##DEBUG_VALUE: StripInvalidUtf8Glyphs:idx <- 0
	##DEBUG_VALUE: Utf8Consume<&node::Skip, <lambda at ../src/util.cc:184:19> >:length <- R15
	##DEBUG_VALUE: Utf8Consume<&node::Skip, <lambda at ../src/util.cc:184:19> >:idx <- 0
	##DEBUG_VALUE: advance <- 0
	##DEBUG_VALUE: glyph <- R10D
	##DEBUG_VALUE: clz:xs <- EBX
	##DEBUG_VALUE: clz:out <- 0
	.loc	1 24 0                  ## ../src/util.cc:24:0
	leal	31(%rax), %ecx
	movl	%ebx, %edx
                                        ## kill: CL<def> CL<kill> ECX<kill>
	shrl	%cl, %edx
	decl	%eax
	testl	%edx, %edx
	jne	LBB0_4
Ltmp19:
## BB#5:                                ## %_ZL3clzj.exit.i
                                        ##   in Loop: Header=BB0_3 Depth=1
	##DEBUG_VALUE: StripInvalidUtf8Glyphs:input <- R14
	##DEBUG_VALUE: StripInvalidUtf8Glyphs:size <- R15
	##DEBUG_VALUE: StripInvalidUtf8Glyphs:idx <- 0
	##DEBUG_VALUE: Utf8Consume<&node::Skip, <lambda at ../src/util.cc:184:19> >:length <- R15
	##DEBUG_VALUE: Utf8Consume<&node::Skip, <lambda at ../src/util.cc:184:19> >:idx <- 0
	##DEBUG_VALUE: advance <- 0
	##DEBUG_VALUE: glyph <- R10D
	.loc	1 131 0                 ## ../src/util.cc:131:0
	notl	%eax
	movzbl	%al, %eax
```

This is on OSX 10.10, clang++ Apple LLVM version 6.0 (clang-600.0.57) (based on LLVM 3.5svn), built with:

```shell
$ c++ '-D_DARWIN_USE_64_BIT_INODE=1' '-DNODE_ARCH="x64"' '-DNODE_TAG=""' '-DNODE_V8_OPTIONS=""' '-DNODE_WANT_INTERNALS=1' '-DHAVE_OPENSSL=1' '-DHAVE_DTRACE=1' '-D__POSIX__' '-DNODE_PLATFORM="darwin"' '-DHTTP_PARSER_STRICT=0' '-D_LARGEFILE_SOURCE' '-D_FILE_OFFSET_BITS=64' -I../src -I../tools/msvs/genfiles -I../deps/uv/src/ares -I/Users/chris/projects/iojs/out/Release/obj/gen -I../deps/cares/include -I../deps/openssl/openssl/include -I../deps/v8/include -I../deps/zlib -I../deps/http_parser -I../deps/uv/include  -Os -gdwarf-2 -mmacosx-version-min=10.5 -arch x86_64 -Wall -Wendif-labels -W -Wno-unused-parameter -std=gnu++0x -fno-rtti -fno-exceptions -fno-threadsafe-statics -fno-strict-aliasing -MMD -MF /Users/chris/projects/iojs/out/Release/.deps//Users/chris/projects/iojs/out/Release/obj.target/iojs/src/util.o.d.raw -S  -c -o /Users/chris/projects/iojs/util.s ../src/util.cc
```